### ISSUE TITLE
Discussion on DyT: Impact on Accuracy, Convergence Speed, and Generalization

### ISSUE THREAD

**Issue by Sweetest316** [jiachenzhu/DyT#7]
Hello, author,

Thank you for sharing your work—I'm very interested in it. I have a few questions:

1、From my understanding, the accuracy improvements in Transformer models mainly stem from the self-attention mechanism. Since normalization layers constitute only a small portion of the model, replacing them with DyT might not fully demonstrate its actual capabilities. Or, would it be more accurate to say that normalization layers and DyT primarily enhance the model’s convergence speed while having a smaller impact on accuracy?

2、Have you conducted experiments on lightweight models to further validate DyT’s generalization ability?

Additionally, does DyT significantly improve the model’s convergence speed?

Looking forward to your insights!

**Comment by Cyril9227** [comment]
For what is worth, I tried to convert some lighter CNNs (`torchvision.models.mobilenet_v2(weights=None)` and `torchvision.models.efficientnet_b0(weights=None)`) using 

```
def convert_ln_to_dyt(module):
    module_output = module
    if isinstance(module, nn.LayerNorm):
        module_output = DynamicTanh(module.normalized_shape, not isinstance(module, LayerNorm2d))
    if isinstance(module, nn.BatchNorm2d):
        normalized_shape = module.num_features
        module_output = DynamicTanh(module.num_features, False)
    for name, child in module.named_children():
        module_output.add_module(name, convert_ln_to_dyt(child))
    del module
    return module_output
```

Both failed to converge on toy datasets (`torchvision.datasets.Imagenette` and `torchvision.datasets.FashionMNIST`) while baseline models reached decent accuracy quickly. Anecdotally, comparing inference speed using `%timeit model(imgs)` showed that the converted models were slightly slower on my machine, probably because the native torch `tanh` implementation is actually expensive (?)

Implementation might be wrong or it might not be a direct plug-in replacement for smaller 2D CNNs depending on BatchNorm vs Transformers with LayerNorm.

**Comment by jiachenzhu** [comment]
Hey, @Cyril9227,

Please take a look at Appendix C in our original paper. We actually found that DyT could not successfully replace BN in ResNet or VGG models, which aligns with your observations.

On the other hand, we found that DyT could effectively replace LN in ConvNeXt models. I suspect this is because DyT is more similar to LN but less similar to BN. So, if you're considering trying lightweight convolutional models, you might want to explore smaller ConvNeXt models.

(In our own experiments, we found that LN also cannot replace BN for those ConvNets. So it somewhat shows that BN works differently to LN or DyT)

**Comment by jiachenzhu** [comment]
Hey, @Sweetest316 

I have a different perspective from your statement. If that were the case, activation functions like GELU or ReLU, which have no trainable parameters, should have minimal impact on accuracy — but we know that’s not true.

Our paper clearly shows that normalization layers exhibit strong non-linearity in such cases. This suggests that they are likely important not only for convergence speed but also for overall accuracy.

That said, we haven't tested DyT on lightweight models to further validate its generalization ability. Do you have any suggestions for lightweight models we could try? We could include additional experiments in our revised version.

In our experiments, it show that DyT has a similar convergence speed to LN layers in most cases.

**Comment by Sweetest316** [comment]
I’m very happy to receive your reply, and I now strongly agree that normalization layers exhibit strong non-linearity.

For lightweight experiments, would it be possible to try commonly used lightweight models such as MobileNet, EfficientNet, or ShuffleNet? These might help accelerate convergence or improve metrics, which could be very beneficial for deployment or downstream tasks.

Looking forward to your response!
